Update release notes for TensorFlow 1.15.0

tensorflow-jenkins opened the pull request:
This PR is intentionally incomplete. One of the Release Owners for 1.15.0
needs to fill in the internal release notes for this version before the PR gets
submitted. Click on the :pencil2: icon in the header for `RELEASE.md` under
"Files Changed" above.

tensorflow-jenkins committed:
Insert release notes place-fill

goldiegadde committed:
Update RELEASE.md

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+# Release 1.15.0
+
```
Let's add here: 

>This is the last 1.x release for TensorFlow. We do not expect to update the 1.x branch with features, although we will issue patch releases to fix vulnerabilities for at least one year. 
>
>TensorFlow 1.15 contains a complete implementation of the 2.0 API in its `compat.v2` module. It contains a copy of the 1.15 main module (without `contrib`) in the `compat.v1` module. TensorFlow 1.15 is able to emulate 2.0 behavior using the `enable_v2_behavior()` function.
>
>This enables writing forward compatible code: by explicitly importing either `tensorflow.compat.v1` or `tensorflow.compat.v2`, you can ensure that your code works without modifications against an installation of 1.15 or 2.0.

goldiegadde replied:
Done!

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Adds `tf.enable_control_flow_v2()` and `tf.disable_control_flow_v2()`.
+* EagerTensor now support buffer interface for tensors.
+* This change bumps the version number of the FullyConnected Op to 5.
+* tensorflow : crash when pointer become nullptr.
```
What is this from? It's useless in its current form. Remove or find the CL.

goldiegadde replied:
Done!

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+
+## Bug Fixes and Other Changes
+
+* Promoting `unbatch` from experimental to core API.
```
that a dataset change? We should be more specific which `unbatch`.

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+## Bug Fixes and Other Changes
+
+* Promoting `unbatch` from experimental to core API.
+* Adds `tf.enable_control_flow_v2()` and `tf.disable_control_flow_v2()`.
```
remove here, covered by the section proposed in major features.

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+# Release 1.15.0
+
+## Major Features and Improvements
+
```
Add here: 


>TensorFlow 1.15 contains a complete implementation of the 2.0 API in its `compat.v2` module. It contains a copy of the 1.15 main module (without `contrib`) in the `compat.v1` module. TensorFlow 1.15 is able to emulate 2.0 behavior using the `enable_v2_behavior()` function.
>
>This enables writing forward compatible code: by explicitly importing either `tensorflow.compat.v1` or `tensorflow.compat.v2`, you can ensure that your code works without modifications against an installation of 1.15 or 2.0.

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+
+* Promoting `unbatch` from experimental to core API.
+* Adds `tf.enable_control_flow_v2()` and `tf.disable_control_flow_v2()`.
+* EagerTensor now support buffer interface for tensors.
```
If we add an example that makes clear what that means, we should move this to major features.

goldiegadde replied:
Done. Checked with alextp@ and added "numpy buffer interface".

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Promoting `unbatch` from experimental to core API.
+* Adds `tf.enable_control_flow_v2()` and `tf.disable_control_flow_v2()`.
+* EagerTensor now support buffer interface for tensors.
+* This change bumps the version number of the FullyConnected Op to 5.
```
Delete.

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* tensorflow : crash when pointer become nullptr.
+* `tf.keras.estimator.model_to_estimator` now supports exporting to tf.train.Checkpoint format, which allows the saved checkpoints to be compatible with `model.load_weights`.
+* Add support for defaulting the value of `cycle_length` argument of `tf.data.Dataset.interleave` to the number of schedulable CPU cores.
+* parallel_for: Add converter for `MatrixDiag`.
```
`parallel_for`

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* parallel_for: Add converter for `MatrixDiag`.
+* Saving a Keras Model using `tf.saved_model.save` now saves the list of variables, trainable variables, regularization losses, and the call function.
+* Add `narrow_range` attribute to `QuantizeAndDequantizeV2` and V3.
+* `OMP_NUM_THREADS` is no longer used by the default Keras config. To configure the number of threads, use `tf.config.threading` APIs.
```
Move this to breaking changes.

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Add `narrow_range` attribute to `QuantizeAndDequantizeV2` and V3.
+* `OMP_NUM_THREADS` is no longer used by the default Keras config. To configure the number of threads, use `tf.config.threading` APIs.
+* Added new op: `tf.strings.unsorted_segment_join`.
+* `tf.keras.model.save_model` and `model.save` now defaults to saving a TensorFlow SavedModel. A SavedModel code is only applicable to `tf.keras`.
```
I don't understand the last sentence in this bullet.

Should this be in breaking changes?

goldiegadde replied:
Added to breaking changes, and removed the last line.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* `OMP_NUM_THREADS` is no longer used by the default Keras config. To configure the number of threads, use `tf.config.threading` APIs.
+* Added new op: `tf.strings.unsorted_segment_join`.
+* `tf.keras.model.save_model` and `model.save` now defaults to saving a TensorFlow SavedModel. A SavedModel code is only applicable to `tf.keras`.
+* Tensorflow code now produces 2 different pip packages: tensorflow_core containing all the code (in the future it will contain only the private implementation) and tensorflow which is a virtual pip package doing forwarding to tensorflow_core (and in the future will contain only the public API of tensorflow)
```
Move this to breaking changes. Add:

>We don't expect this to be breaking, unless you were importing directly from the implementation.

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Deprecated `tf.keras.experimental.export_saved_model` and `tf.keras.experimental.function`. Please use `tf.keras.models.save_model(..., save_format='tf')` and `tf.keras.models.load_model` instead.
+* Add new `TypeSpec` classes
+* CloudBigtable version updated to v0.10.0 BEGIN_PUBLIC CloudBigtable version updated to v0.10.0
+* Deprecated the use of `constraint=` and `.constraint` with ResourceVariable.
```
move to breaking changes

goldiegadde replied:
Done

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Add new `TypeSpec` classes
+* CloudBigtable version updated to v0.10.0 BEGIN_PUBLIC CloudBigtable version updated to v0.10.0
+* Deprecated the use of `constraint=` and `.constraint` with ResourceVariable.
+* Expose Head as public API.
```
`Head`, ideally, use fully qualified name.

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* CloudBigtable version updated to v0.10.0 BEGIN_PUBLIC CloudBigtable version updated to v0.10.0
+* Deprecated the use of `constraint=` and `.constraint` with ResourceVariable.
+* Expose Head as public API.
+* AutoGraph is now applied automatically to user functions passed to APIs of `tf.data` and `tf.distribute`. If AutoGraph is disabled in the the calling code, it will also be disabled in the user functions.
```
Delete. This is covered by a different bullet.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Update docstring for gather to properly describe the non-empty batch_dims case.
+* Added `tf.sparse.from_dense` utility function.
+* Add `GATHER` support to NN API delegate
+* AutoGraph translates Python control flow into TensorFlow expressions, allowing users to write regular Python inside `tf.function`-decorated functions. AutoGraph is also applied in functions used with `tf.data`, `tf.distribute` and `tf.keras` APIs.
```
Move this to major changes.

goldiegadde replied:
Done

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+## Major Features and Improvements
+
+## Breaking Changes
+
```
Add here (@gunan to confirm words are correct): 

>TensorFlow 1.15 is built using devtoolset7 on Ubuntu 16. This may lead to ABI incompatibilities with extensions built against earlier versions of TensorFlow.

goldiegadde replied:
Updated. @gunan PTAL

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* AutoGraph translates Python control flow into TensorFlow expressions, allowing users to write regular Python inside `tf.function`-decorated functions. AutoGraph is also applied in functions used with `tf.data`, `tf.distribute` and `tf.keras` APIs.
+* Improved ragged tensor support in `TensorFlowTestCase`.
+* Makes the a-normal form transformation in Pyct configurable as to which nodes are converted to variables and which are not.
+* `keras.backend.resize_images` (and consequently, `keras.layers.Upsampling2D`) behavior has changed, a bug in the resizing implementation was fixed.
```
Move to breaking changes

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Makes the a-normal form transformation in Pyct configurable as to which nodes are converted to variables and which are not.
+* `keras.backend.resize_images` (and consequently, `keras.layers.Upsampling2D`) behavior has changed, a bug in the resizing implementation was fixed.
+* `ResizeInputTensor` now works for all delegates
+* Start of open development of TF, TFLite, XLA MLIR dialects.
```
remove this. It's nice, but doesn't need to be part of release notes.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Add `tf.ragged.stack`.
+* Add delegate support for `QUANTIZED_16BIT_LSTM`.
+* Add an `implementation=3` mode for `tf.keras.layers.LocallyConnected2D` and `tf.keras.layers.LocallyConnected1D` layers using `tf.SparseTensor` to store weights,  allowing a dramatic speedup for large sparse models.
+* `tf.while_loop` emits a StatelessWhile op if the cond and body functions are stateless and do not touch any resources.
```
would be nice to group these better: this one should go next to the corresponding tf.cond change above.

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Add an `implementation=3` mode for `tf.keras.layers.LocallyConnected2D` and `tf.keras.layers.LocallyConnected1D` layers using `tf.SparseTensor` to store weights,  allowing a dramatic speedup for large sparse models.
+* `tf.while_loop` emits a StatelessWhile op if the cond and body functions are stateless and do not touch any resources.
+* Refactors code in Quant8 LSTM support to reduce TFLite binary size.
+* Layers now default to float32, and automatically cast their inputs to the layer's dtype. If you had a model that used float64, it will probably silently use float32 in TensorFlow2, and a warning will be issued that starts with Layer "layer-name" is casting an input tensor from dtype float64 to the layer's dtype of float32. To fix, either set the default dtype to float64 with `tf.keras.backend.s ...
```
Move to breaking changes.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Fix memory allocation problem when calling `AddNewInputConstantTensor`.
+* Delegate application failure leaves interpreter in valid state.
+* Enable the Keras compile API `experimental_run_tf_function` flag by default. This flag enables single training/eval/predict execution path. With this 1. All input types are converted to `Dataset`. 2. When distribution strategy is not specified this goes through the no-op distribution strategy path. 3. Execution is wrapped in tf.function unless `run_eagerly=True` is set in compile.
+* `tf.cond`, `tf.while` and if and while in AutoGraph now accept a nonscalar predicate if has a single element. This does not affec non-V2 control flow.
```
`if` and `while` and "affect".

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Added support for `FusedBatchNormV3` in converter.
+* A ragged to dense op for directly calculating tensors.
+* Converts hardswish subgraphs into atomic ops.
+* The equality operation on Tensors & Variables now compares on value instead of id(). As a result, both Tensors & Variables are no longer hashable types.
```
Move to major features, and rewrite to:

> Adds `enable_tensor_equality()`, which switches the behavior such that 
> 
> * `Tensor`s are no longer hashable
> * `Tensor`s can be compared with `==` and `!=`, yielding a Boolean `Tensor` with element-wise comparison results.
>
> This will be the default behavior in 2.0

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Converts hardswish subgraphs into atomic ops.
+* The equality operation on Tensors & Variables now compares on value instead of id(). As a result, both Tensors & Variables are no longer hashable types.
+* Raise error if `batch_size` argument is used when input is dataset/generator/keras sequence.
+* The equality operation on Tensors & Variables now compares on value instead of id(). As a result, both Tensors & Variables are no longer hashable types.
```
Remove, this is a duplicate.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* The equality operation on Tensors & Variables now compares on value instead of id(). As a result, both Tensors & Variables are no longer hashable types.
+* Raise error if `batch_size` argument is used when input is dataset/generator/keras sequence.
+* The equality operation on Tensors & Variables now compares on value instead of id(). As a result, both Tensors & Variables are no longer hashable types.
+* Some `tf.assert_*` methods now raise assertions at operation creation time (i.e. when this Python line executes) if the input tensors' values are known at that time, not during the session.run(). When this happens, a noop is returned and the input tensors are marked non-feedable. In other words, if they are used as keys in `feed_dict` argument to session.run(), an error will be raised. Also, be ...
```
Move to breaking changes

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,83 @@
+* Raise error if `batch_size` argument is used when input is dataset/generator/keras sequence.
+* The equality operation on Tensors & Variables now compares on value instead of id(). As a result, both Tensors & Variables are no longer hashable types.
+* Some `tf.assert_*` methods now raise assertions at operation creation time (i.e. when this Python line executes) if the input tensors' values are known at that time, not during the session.run(). When this happens, a noop is returned and the input tensors are marked non-feedable. In other words, if they are used as keys in `feed_dict` argument to session.run(), an error will be raised. Also, be ...
+* The equality operation on Tensors & Variables now compares on value instead of id(). As a result, both Tensors & Variables are no longer hashable types.
```
remove duplicate

goldiegadde committed:
Update RELEASE.md

goldiegadde committed:
Update RELEASE.md

goldiegadde committed:
Update RELEASE.md

goldiegadde committed:
Update RELEASE.md

gunan commented on RELEASE.md:
```diff
@@ -1,3 +1,86 @@
+
+## Breaking Changes
+* Tensorflow code now produces 2 different pip packages: tensorflow_core containing all the code (in the future it will contain only the private implementation) and tensorflow which is a virtual pip package doing forwarding to tensorflow_core (and in the future will contain only the public API of tensorflow). We don't expect this to be breaking, unless you were importing directly from the impleme ...
+* TensorFlow 1.15 is built using devtoolset7 on Ubuntu 16. This may lead to ABI incompatibilities with extensions built against earlier versions of TensorFlow.
```
Let's add GCC7 in parenthesis, just to make sure it is explicitly mentioned.

goldiegadde committed:
Update RELEASE.md

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,86 @@
+This is the last 1.x release for TensorFlow. We do not expect to update the 1.x branch with features, although we will issue patch releases to fix vulnerabilities for at least one year. 
+
+## Major Features and Improvements
+* TensorFlow 1.15 contains a complete implementation of the 2.0 API in its compat.v2 module. It contains a copy of the 1.15 main module (without contrib) in the compat.v1 module. TensorFlow 1.15 is able to emulate 2.0 behavior using the enable_v2_behavior() function.
```
`contrib`, `compat.v1`, `enable_v2_behavior()`

goldiegadde replied:
Done

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,86 @@
+
+## Major Features and Improvements
+* TensorFlow 1.15 contains a complete implementation of the 2.0 API in its compat.v2 module. It contains a copy of the 1.15 main module (without contrib) in the compat.v1 module. TensorFlow 1.15 is able to emulate 2.0 behavior using the enable_v2_behavior() function.
+This enables writing forward compatible code: by explicitly importing either tensorflow.compat.v1 or tensorflow.compat.v2, you can ensure that your code works without modifications against an installation of 1.15 or 2.0.
```
`tensorflow.compat.v1`, `tensorflow.compat.v2`

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,86 @@
+* Enable v2 control flow as part of `tf.enable_v2_behavior()` and `TF2_BEHAVIOR=1`.
+* AutoGraph translates Python control flow into TensorFlow expressions, allowing users to write regular Python inside `tf.function`-decorated functions. AutoGraph is also applied in functions used with `tf.data`, `tf.distribute` and `tf.keras` APIs.
+* Adds enable_tensor_equality(), which switches the behavior such that: 
+  Tensors are no longer hashable
```
Make a bullet list

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,86 @@
+
+## Breaking Changes
+* Tensorflow code now produces 2 different pip packages: tensorflow_core containing all the code (in the future it will contain only the private implementation) and tensorflow which is a virtual pip package doing forwarding to tensorflow_core (and in the future will contain only the public API of tensorflow). We don't expect this to be breaking, unless you were importing directly from the impleme ...
+* TensorFlow 1.15 is built using devtoolset7(GCC7) on Ubuntu 16. This may lead to ABI incompatibilities with extensions built against earlier versions of TensorFlow.
```
space before (

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,86 @@
+  * `OMP_NUM_THREADS` is no longer used by the default Keras config. To configure the number of threads, use `tf.config.threading` APIs.
+  * `tf.keras.model.save_model` and `model.save` now defaults to saving a TensorFlow SavedModel.
+  * `keras.backend.resize_images` (and consequently, `keras.layers.Upsampling2D`) behavior has changed, a bug in the resizing implementation was fixed.
+  * Layers now default to float32, and automatically cast their inputs to the layer's dtype. If you had a model that used float64, it will probably silently use float32 in TensorFlow2, and a warning will be issued that starts with Layer "layer-name" is casting an input tensor from dtype float64 to the layer's dtype of float32. To fix, either set the default dtype to float64 with `tf.keras.backend ...
```
`float32` and `float64`

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,86 @@
+* Added new op: `tf.strings.unsorted_segment_join`.
+* Add HW acceleration support for `topK_v2`.
+* Add new `TypeSpec` classes.
+* CloudBigtable version updated to v0.10.0 BEGIN_PUBLIC CloudBigtable version updated to v0.10.0
```
some bad formatting here

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,86 @@
+* Add new `TypeSpec` classes.
+* CloudBigtable version updated to v0.10.0 BEGIN_PUBLIC CloudBigtable version updated to v0.10.0
+* Expose `Head` as public API.
+* Update docstring for gather to properly describe the non-empty batch_dims case.
```
`batch_dims`

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,86 @@
+* `tf.while_loop` emits a StatelessWhile op if the cond and body functions are stateless and do not touch any resources.
+* Refactors code in Quant8 LSTM support to reduce TFLite binary size.
+* Add support of local soft device placement for eager op.
+* Pass partial_pivoting to the `_TridiagonalSolveGrad`.
```
I'd remove this one, it's too obscure.

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,86 @@
+* Add support of local soft device placement for eager op.
+* Pass partial_pivoting to the `_TridiagonalSolveGrad`.
+* Add HW acceleration support for `LogSoftMax`.
+* Added a function nested_value_rowids for ragged tensors.
```
`nested_value_rowids`

goldiegadde replied:
Done.

martinwicke commented on RELEASE.md:
```diff
@@ -1,3 +1,86 @@
+* Pass partial_pivoting to the `_TridiagonalSolveGrad`.
+* Add HW acceleration support for `LogSoftMax`.
+* Added a function nested_value_rowids for ragged tensors.
+* fixed a bug in histogram_op.cc.
```
remove this, it's too obscure.

goldiegadde replied:
Done.

goldiegadde committed:
Update RELEASE.md

goldiegadde committed:
Update RELEASE.md

goldiegadde committed:
Update RELEASE.md

goldiegadde committed:
Update RELEASE.md

goldiegadde committed:
Update RELEASE.md

yifeif committed:
Add pip consolidation to 1.15.0 release note.

mihaimaruseac committed:
Add `tf.estimator` release notes

At one point in the future we might want to not add those notes into TF's ones to properly separate the release cycles of the two projects.

But for now they are still entangled

mihaimaruseac commented:
I added the `tf.estimator` notes, please take a look.

angerson commented on RELEASE.md:
```diff
@@ -1,3 +1,90 @@
+
+## Major Features and Improvements
+* As [announced](https://groups.google.com/a/tensorflow.org/forum/#!topic/developers/iRCt5m4qUz0), `tensorflow` pip package will by default include GPU support (same as `tensorflow-gpu` now) for the platforms we currently have GPU support (Linux and Windows). It will work on machines with and without Nvidia GPUs. `tensorflow-gpu` will still be available, and CPU-only packages can be downloaded at ...
+* TensorFlow 1.15 contains a complete implementation of the 2.0 API in its `compat.v2 module`. It contains a copy of the 1.15 main module (without `contrib`) in the `compat.v1 module`. TensorFlow 1.15 is able to emulate 2.0 behavior using the `enable_v2_behavior()` function.
```
This looks like it should be "`compat.v2` module"

goldiegadde replied:
Done.

mihaimaruseac commented on RELEASE.md:
```diff
@@ -1,3 +1,90 @@
+* AutoGraph translates Python control flow into TensorFlow expressions, allowing users to write regular Python inside `tf.function`-decorated functions. AutoGraph is also applied in functions used with `tf.data`, `tf.distribute` and `tf.keras` APIS.
+* Adds `enable_tensor_equality()`, which switches the behavior such that: 
+  * Tensors are no longer hashable.
+  * Tensors can be compared with == and !=, yielding a Boolean Tensor with element-wise comparison results. This will be the default behavior in 2.0.
```
Nit: `==` and `!=`

goldiegadde replied:
Done.

angerson commented on RELEASE.md:
```diff
@@ -1,3 +1,90 @@
+  * Tensors can be compared with == and !=, yielding a Boolean Tensor with element-wise comparison results. This will be the default behavior in 2.0.
+
+## Breaking Changes
+* Tensorflow code now produces 2 different pip packages: tensorflow_core containing all the code (in the future it will contain only the private implementation) and tensorflow which is a virtual pip package doing forwarding to tensorflow_core (and in the future will contain only the public API of tensorflow). We don't expect this to be breaking, unless you were importing directly from the impleme ...
```
Nit: `tensorflow_core` and `tensorflow` should have backticks.

goldiegadde replied:
Done.

angerson reviewed (commented):
I added a few small nitpicks, but overall, this looks good.

goldiegadde committed:
Update RELEASE.md
